abd-clam/src/cakes/knn/sieve_v1.rs

```diff
@@ -29,257 +27,154 @@ where
-impl<'a, T: Send + Sync + Copy, U: Number, D: Dataset<T, U>> SieveV1<'a, T, U, D> {
-    /// Creates a new instance of a `Sieve V1`.
-    pub fn new(tree: &'a Tree<T, U, D>, query: T, k: usize) -> Self {
-        Self {
-            tree,
-            query,
-            k,
-            grains: Vec::new(),
-            is_refined: false,
-            hits: Hits::default(),
+    let data = tree.data();
+    let c = tree.root();
+    let d = c.distance_to_instance(data, query);
+
+    let mut grains = vec![Grain::new(c, d)];
+    let mut hits = Hits::new(k);
+
+    loop {
+        // The threshold is the minimum distance, so far, which guarantees that
+        // the k nearest neighbors are within the threshold.
+        let i = Grain::partition(&mut grains, k, hits.len());
+        let threshold = grains[i].d;
+
+        // Remove hits which are outside the threshold.
+        hits.pop_until(threshold);
+
+        let (insiders, non_insiders) = grains.split_at_mut(i);
+        if non_insiders.is_empty() {
+            grains_to_hits(data, query, &mut hits, insiders);
+            break;
         }
-    }
-
-    /// One-time computation of `grains.`
-    pub fn initialize_grains(&mut self) {
-        let root = self.tree.root();
-        let distance = root.distance_to_instance(self.tree.data(), self.query);
-        let grain = Grain::new(root, distance + root.radius, root.cardinality);
-        self.grains = vec![grain];
-    }
-
-    /// Returns whether search is complete, i.e., whether hits
-    /// contains the k-nearest neighbors.
-    const fn is_refined(&self) -> bool {
-        self.is_refined
-    }
 
-    /// One iteration of the refinement step.
-    pub fn refine_step(&mut self) {
-        // Determines the index of the grain such that if we only retain that grain
-        // and every closer grain, we are guaranteed to have at least k points.
-        // Threshold is the distance of the furthest point in the grain.
-        let i = Grain::partition_kth(&mut self.grains, self.k);
-        let ith_grain = &self.grains[i];
-        let threshold = ith_grain.d;
-
-        // Filters hits by being outside the threshold.
-        while !self.hits.is_empty()
-            && self
-                .hits
-                .peek_max()
-                .unwrap_or_else(|| unreachable!("`hits` is non-empty"))
-                .1
-                .number
-                > threshold
-        {
-            self.hits
-                .pop_max()
-                .unwrap_or_else(|| unreachable!("`hits` is non-empty"));
+        let (small_insiders, insiders) = insiders
+            .iter_mut()
+            .map(|&mut g| g)
+            .partition::<Vec<_>, _>(|g| g.is_small(k));
+        grains_to_hits(data, query, &mut hits, &small_insiders);
+
+        let (leaves, non_leaves) = non_insiders
+            .iter_mut()
+            .filter(|g| !g.is_outside(threshold))
+            .map(|&mut g| g)
+            .partition::<Vec<_>, _>(|g| g.is_leaf);
+
```

Comment:
See what happens if you use is_small() rather than is_leaf() here. Likely makes negligible difference but now that we are rewarded for having more points in hits, it may offer improvements